platform-sdk/docs/proposals/prehandle-buffer/prehandle-buffer-proposal.md

```diff
@@ -0,0 +1,183 @@
+# Pre-handle buffer design proposal
+
+---
+
+## Summary
+
+Proposal to replace the custom thread synchronization solution for pre-handle with a component that would buffer rounds
+until pre-handle is complete.
+
+| Metadata           | Entities          | 
+|--------------------|-------------------|
+| Designers          | Lazar             |
+| Functional Impacts | Platform internal |
+| Related Proposals  | none              |
+| HIPS               | none              |
+
+---
+
+## Purpose and Context
+
+Currently, there is a CountdownLatch inside each event which ensure that pre-handle is always done before the event is
+handled in consensus order. The main advantage of this solution is that it is a very simple implementation. But there 
+are more drawbacks which are:
+
+- It can lead to performance inefficiencies because it has the possibility of blocking the event handling thread.
+- It mixes business logic with concurrency control, which is against the established architecture.
+- It creates a possibility for a unit test that to hang indefinitely.
+- It is not compatible with TURTLE tests for the same reason, it could make a test hang indefinitely.
+
+The proposal is to replace this mechanism with a component that would buffer rounds until pre-handle is complete.
```

Comment:
We don't really have a separate thread for handling, it ends up just being a thread from the fork join pool. The performance problem (according to Oleg) is that blocking and unblocking a thread is very inefficient. It is better to ensure that a round is ready for handling and then give it to a thread to perform the work.

Reply:
Thanks for the clarification. I'm unsure if it's a great idea to reuse a single pool for unrelated tasks, but I presume this is the current threading model.

ForkJoinPools provide the `ManagedBlocker` feature designed specifically to solve the problem of blocked threads in the pool. Could we explore pros and cons of this alternative solution, too, and see if perhaps it could introduce less extra complexity?

Another alternative solution could involve modifying the method that feeds individual events to the pool. If we assume that currently it looks something like this (pseudo code):

```
round.getEvents().parallelStream().forEach(pool::submit);
```

then it could be complicated slightly as follows:

```
set = new HashSet(round.getEvents());
while (!set.isEmpty()) {
   list = set.parallelStream().filter(e -> e.isPrehandled()).peek(pool::submit).toList();
   set.removeAll(list);
   sleep(1); // optional
}
```

This way we avoid submitting potentially blocking events to the pool, and we also avoid introducing any new components into the system. We also avoid introducing any buffers spanning multiple rounds and instead use a "local", round-specific buffer in the form of that `set` above, thus avoiding any additional links to the health monitor.

Could we explore pros and cons of this solution as well?

Reply:
1. I understand the intent. However, I don't find this being a good reason to avoid using a machinery provided by the JDK out of the box to solve the exact problem that we're having. I'm okay if you want to build some wrappers or abstractions on top of that machinery in order to make it look as if the business logic is unaware of the concurrency (provided, it's decided that it's reasonable to allocate capacity on such an effort.) However, regardless of having those extra abstractions, I believe that as long as we use a shared pool, then at the lowest level of our implementation, the ManagedBlockers would be the only reasonable way to solve the problem.
2. To rephrase the second option, the idea is to check if an event has been pre-handled before submitting a task that handles the event into the ForkJoinPool, rather than perform this check in the task implementation itself after it's been submitted to the pool already.